app/client/lib/trapTabKey.ts

```diff
@@ -0,0 +1,244 @@
+  if (activeEl.shadowRoot)
+    { return getActiveEl(activeEl.shadowRoot) || document.activeElement; }
+
+  // If not, we can just return the active element
+  return activeEl;
+}
+
+/**
+ * Determine if an element can have focusable children. Useful for bailing out
+ * early when walking the DOM tree.
+ * @example
+ * This div is inert, so none of its children can be focused, even though they
+ * meet our criteria for what is focusable. Once we check the div, we can skip
+ * the rest of the subtree.
+ * ```html
+ * <div inert>
+ *   <button>Button</button>
+ *   <a href="#">Link</a>
+ * </div>
+ * ```
+ */
+function canHaveFocusableChildren(el: HTMLElement) {
+  // The browser will never send focus into a Shadow DOM if the host element
+  // has a negative tabindex. This applies to both slotted Light DOM Shadow DOM
+  // children
+  if (el.shadowRoot && el.getAttribute('tabindex') === '-1') { return false; }
+
+  // Elemments matching this selector are either hidden entirely from the user,
+  // or are visible but unavailable for interaction. Their descentants can never
+  // receive focus.
+  return !el.matches(':disabled,[hidden],[inert]');
+}
+
+
+function getNextChildEl(el: ParentNode, forward: boolean) {
+  return forward ? el.firstElementChild : el.lastElementChild;
+}
+
+function getNextSiblingEl(el: HTMLElement, forward: boolean) {
+  return forward ? el.nextElementSibling : el.previousElementSibling;
+}
+
+/**
+ * Determine if an element is hidden from the user.
+ */
+const isHidden = (el: HTMLElement) => {
+  // Browsers hide all non-<summary> descendants of closed <details> elements
+  // from user interaction, but those non-<summary> elements may still match our
+  // focusable-selectors and may still have dimensions, so we need a special
+  // case to ignore them.
+  if (
+    el.matches('details:not([open]) *') &&
+    !el.matches('details>summary:first-of-type')
+  )
+    { return true; }
+
+  // If this element has no painted dimensions, it's hidden.
+  return !(el.offsetWidth || el.offsetHeight || el.getClientRects().length);
+};
+
+/**
+ * Determine if an element is focusable and has user-visible painted dimensions.
+ */
+const isFocusable = (el: HTMLElement) => {
+  // A shadow host that delegates focus will never directly receive focus,
+  // even with `tabindex=0`. Consider our <fancy-button> custom element, which
+  // delegates focus to its shadow button:
+  //
+  // <fancy-button tabindex="0">
+  //  #shadow-root
+  //  <button><slot></slot></button>
+  // </fancy-button>
+  //
+  // The browser acts as as if there is only one focusable element – the shadow
+  // button. Our library should behave the same way.
+  if (el.shadowRoot?.delegatesFocus) { return false; }
+
+  return el.matches(focusableSelectors.join(',')) && !isHidden(el);
+};
+
+const notInert = ':not([inert]):not([inert] *)';
+const notNegTabIndex = ':not([tabindex^="-"])';
+const notDisabled = ':not(:disabled)';
+
+const focusableSelectors = [
```

Comment:
That part specifically is available under the [focusable-selectors](https://github.com/KittyGiraudel/focusable-selectors) package if you’d like. Very easy and portable and says what’s on the tin. :)